### DISCUSSION TITLE
classes/class_characterbody2d

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#160] [User-contributed notes]
# classes/class_characterbody2d

Inherits: PhysicsBody2D< CollisionObject2D< Node2D< CanvasItem< Node< Object A 2D physics body specialized for characters moved by script. Description: CharacterBody2D is a specialized class for ph...

https://docs.godotengine.org/en/stable/classes/class_characterbody2d.html

**Comment by lemonhall** [comment]
remember to use this,for collision

```gdscript
func _physics_process(delta):
	velocity = Vector2.ZERO # The player's movement vector.
	if Input.is_action_pressed("ui_up",true):
		animated_sprite_2d.play("up")
		velocity.y = -speed
	if Input.is_action_pressed("ui_down",true):
		animated_sprite_2d.play("face_down")
		velocity.y = speed
	if Input.is_action_pressed("ui_left",true):
		animated_sprite_2d.play("left")
		velocity.x = -speed
	if Input.is_action_pressed("ui_right",true):
		animated_sprite_2d.play("right")
		velocity.x = speed
	if velocity.length() > 0:
		move_and_slide()
	pass	
```

move_and_slide()!!!!!

this takes me 2 hours to work

**Comment by lackhand** [comment]
The docs say "but they affect other physics bodies in their path" -- but experimentally they don't.

If I have a scene with a 2d character controller rectangle, a sphere with some initial velocity and a physics material that's bouncy, and a static body/world boundary colliders -- the sphere does not bounce off of the player as though the player were a paddle. Their collision isn't a momentum preserving interaction. "affect" feels like the wrong word to me?

**Comment by UnTabouret** [comment]
The entry for `get_last_slide_collision` doesn't state that it will return `null` (and not an error) if no collision happened during the last call of `move_and_slide`, I feel like it should be explicitly stated.

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot/pull/112180

**Comment by kassheart** [comment]
the move and slide method description says that it modifies velocity if a slide collision occurs which just isn't true for floating character bodies, as the velocity stays the same no matter what unlike grounded character bodies; there should be a warning for this at the very least
